lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,178 @@
+--
+-- @author Yauheni Kirylau &lt;yawghen@gmail.com&gt;
+-- @copyright 2016-2017 Yauheni Kirylau
+-- @module beautiful.gtk
+---------------------------------------------------------------------------
+local gears_debug = require("gears.debug")
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+local gtk = {
+  cached_theme_variables = nil
+}
+
+
+local function convert_gtk_color_to_hex(gtk_color)
+  local rgb = gtk_color:to_string()
+  local m = rgb:gmatch("[0-9]+")
+  return m and (
+    "#" .. string.format("%2.2x", m()) .. string.format("%2.2x", m()) .. string.format("%2.2x", m())
+  )
+end
+
+
+local function lookup_gtk_color_to_hex(_style_context, color_name)
+  local color = _style_context:lookup_color(color_name)
+  if not color then
+      return nil
+  end
+  local result = convert_gtk_color_to_hex(color)
+  --color:free()
+  return result
+end
+
+local function get_gtk_color_property_to_hex(_style_context, property_name, state)
+    local property = _style_context:get_property(property_name, state)
+    if not property then
+      return nil
+    end
+    local result = convert_gtk_color_to_hex(property.value)
+    property:unset()
+    return result
+end
+
+
+--- Get GTK+3 theme variables from GtkStyleContext
+-- @treturn table subj
+function gtk.get_theme_variables()
+  if gtk.cached_theme_variables then
+    return gtk.cached_theme_variables
+  end
+
+  local result = {}
+  local lgi = require('lgi')
+  local Gtk = lgi.Gtk
+  local window
+  pcall(function()
+    window = Gtk.Window{
+      --on_destroy = Gtk.main_quit,
+    }
+  end)
+  if not window then
+    gears_debug.print_warning("Seems like GTK+3 is not installed or theme is not set correctly.")
+    return nil
+  end
+  --window:set_override_redirect(true)
+  local style_context = window:get_style_context()
+
+  local font = style_context:get_font(Gtk.StateFlags.NORMAL)
+  --result.font_family = font:get_family()
+  --result.font_size = font:get_size()/1024
+  result.font = font:to_string()
+  for result_key, style_context_property in pairs({
+    bg_color="background-color",
+    fg_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        style_context, style_context_property, Gtk.StateFlags.NORMAL
+    )
+  end
+
+  local entry = Gtk.Entry()
+  local entry_style_context = entry:get_style_context()
+  for result_key, style_context_property in pairs({
+    border_radius="border-radius",
+    border_width="border-top-width",
+  }) do
+    local property = entry_style_context:get_property(style_context_property, Gtk.StateFlags.NORMAL)
+    result[result_key] = property.value
+    property:unset()
+  end
+  for result_key, style_context_property in pairs({
+    base_color="background-color",
+    text_color="color",
+  }) do
+    result[result_key] = get_gtk_color_property_to_hex(
+        entry_style_context, style_context_property, Gtk.StateFlags.NORMAL
+    )
+  end
+  entry:destroy()
```

Comment:
I think the docs for this module should say what this actually does... So the `bg/fg_color`s come from the background-color and color of a GtkWindow, the border width is the border-top-width of a GtkEntry and the base_color and text_color come from the background and foreground color of a GtkEntry?!?

Reply:
right, if you take a look on gtk3-widget-factory (or any other gtk3 app) you will see what stylecontext values bg/fg are used for windows, text/base - for entries

however if the same values are defined in stylecontext, stylecontext have a higher prority (because of pixmaps/gradients issue described in the previous comment)